Branch-SDK/src/main/java/io/branch/referral/Branch.java

```diff
@@ -2257,200 +2234,12 @@ private boolean checkIntentForUnusedBranchLink(Intent intent) {
-                                handleFailure(0, status);
-                            }
-                        }
-                        //On Network error or Branch is down fail all the pending requests in the queue except
-                        //for request which need to be replayed on failure.
-                        else {
-                            hasNetwork_ = false;
-                            //Collect all request from the queue which need to be failed.
-                            ArrayList<ServerRequest> requestToFail = new ArrayList<>();
-                            for (int i = 0; i < requestQueue_.getSize(); i++) {
-                                requestToFail.add(requestQueue_.peekAt(i));
-                            }
-                            //Remove the requests from the request queue first
-                            for (ServerRequest req : requestToFail) {
-                                if (req == null || !req.shouldRetryOnFail()) { // Should remove any nullified request object also from queue
-                                    requestQueue_.remove(req);
-                                }
-                            }
-                            // Then, set the network count to zero, indicating that requests can be started again.
-                            networkCount_ = 0;
-
-                            //Finally call the request callback with the error.
-                            for (ServerRequest req : requestToFail) {
-                                if (req != null) {
-                                    req.handleFailure(status, serverResponse.getFailReason());
-                                    //If request need to be replayed, no need for the callbacks
-                                    if (req.shouldRetryOnFail())
-                                        req.clearCallbacks();
-                                }
-                            }
-                        }
-                    }
-                    // If the request succeeded
-                    else {
-                        hasNetwork_ = true;
-                        //On create  new url cache the url.
-                        if (thisReq_ instanceof ServerRequestCreateUrl) {
-                            if (serverResponse.getObject() != null) {
-                                final String url = serverResponse.getObject().getString("url");
-                                // cache the link
-                                linkCache_.put(((ServerRequestCreateUrl) thisReq_).getLinkPost(), url);
-                            }
-                        }
-                        //On Logout clear the link cache and all pending requests
-                        else if (thisReq_ instanceof ServerRequestLogout) {
-                            linkCache_.clear();
-                            requestQueue_.clear();
-                        }
-                        requestQueue_.dequeue();
```

Comment:
I could not understand why the successful requests were being removed from the request queue **by index** and not **by reference** to the successfully completed request, which is just so much safer.